engine/table/src/main/java/io/deephaven/engine/table/impl/locations/impl/AbstractTableLocationProvider.java

```diff
@@ -102,29 +122,128 @@ public final ImmutableTableKey getKey() {
 
     @Override
     protected final void deliverInitialSnapshot(@NotNull final TableLocationProvider.Listener listener) {
-        unmodifiableTableLocationKeys.forEach(listener::handleTableLocationKeyAdded);
+        listener.handleTableLocationKeysUpdate(unmodifiableTableLocationKeys, null);
+    }
+
+    /**
+     * Internal method to begin an atomic transaction of location adds and removes.
+     *
+     * @param token A token to identify the transaction
+     */
+    protected void beginTransaction(@NotNull final Object token) {
+        synchronized (transactionLock) {
+            // Verify that we can start a new transaction with this token.
+            if (transactions.containsKey(token)) {
+                throw new IllegalStateException("A transaction with token " + token + " is currently open.");
+            }
+            transactions.put(token, new Transaction());
+        }
+    }
+
+    /**
+     * Internal method to end an atomic transaction of location adds and removes.
+     *
+     * @param token A token to identify the transaction
+     */
+    protected void endTransaction(@NotNull final Object token) {
+        final Transaction transaction;
+        synchronized (transactionLock) {
+            // Verify that this transaction is open.
+            transaction = transactions.remove(token);
+            if (transaction == null) {
+                throw new IllegalStateException("No transaction with token " + token + " is currently open.");
+            }
+        }
+
+        final Collection<ImmutableTableLocationKey> addedImmutableKeys =
+                new ArrayList<>(transaction.locationsAdded.size());
+        final Collection<ImmutableTableLocationKey> removedImmutableKeys =
+                new ArrayList<>(transaction.locationsRemoved.size());
+
+        // Process the accumulated adds and removes under a lock on `tableLocations` to keep modifications atomic to
+        // other holders of this lock.
+        synchronized (tableLocations) {
+            if (transaction.locationsAdded != EMPTY_TABLE_LOCATION_KEYS
+                    || transaction.locationsRemoved != EMPTY_TABLE_LOCATION_KEYS) {
+                for (TableLocationKey locationKey : transaction.locationsAdded) {
```

Comment:
I think we need to mutate table locations for removes before adds, in case of replace.